langchain/src/chat_models/universal.ts

```diff
@@ -0,0 +1,870 @@
+import {
+  BaseLanguageModelInput,
+  ToolDefinition,
+} from "@langchain/core/language_models/base";
+import {
+  BaseChatModel,
+  BaseChatModelParams,
+  type BaseChatModelCallOptions,
+} from "@langchain/core/language_models/chat_models";
+import { BaseMessage, type AIMessageChunk } from "@langchain/core/messages";
+import {
+  Runnable,
+  type RunnableBatchOptions,
+  RunnableBinding,
+  type RunnableConfig,
+  type RunnableToolLike,
+  ensureConfig,
+} from "@langchain/core/runnables";
+import {
+  AsyncGeneratorWithSetup,
+  IterableReadableStream,
+} from "@langchain/core/utils/stream";
+import {
+  type LogStreamCallbackHandlerInput,
+  type RunLogPatch,
+  type StreamEvent,
+} from "@langchain/core/tracers/log_stream";
+import { type StructuredToolInterface } from "@langchain/core/tools";
+import { CallbackManagerForLLMRun } from "@langchain/core/callbacks/manager";
+import { ChatResult } from "@langchain/core/outputs";
+
+// TODO: remove once `EventStreamCallbackHandlerInput` is exposed in core.
+interface EventStreamCallbackHandlerInput
+  extends Omit<LogStreamCallbackHandlerInput, "_schemaFormat"> {}
+
+const _SUPPORTED_PROVIDERS = [
+  "openai",
+  "anthropic",
+  "azure_openai",
+  "cohere",
+  "google-vertexai",
+  "google-genai",
+  "ollama",
+  "together",
+  "fireworks",
+  "mistralai",
+  "groq",
+  "bedrock",
+] as const;
+
+export type ChatModelProvider = (typeof _SUPPORTED_PROVIDERS)[number];
+
+// This is needed so when `bindTools` is called on a configurable model, the typing is correct.
+abstract class BaseChatModelWithBindTools<
```

Comment:
Do we need this anymore? Can't `_ConfigurableModel` just extend `BaseChatModel` directly?